### PR TITLE
Major update

### PR DESCRIPTION
**Pull request by cfrancesco**
Summary:
* Switch from master to main
* Add pitch modelling
* Remove unnecessary model layers + improved normalization/dropouts
* Autoregressive model specialized as Aligner
* Replaced iterative duration extraction with Dijkstra
* Stratified configuration and data storage to decouple training session and reduce disk space
* Simplified scripts
* Add CLI predictions

### CODE REVIEW

File: utils/logging_utils.py [hunk 1]
```diff
@@ -109,14 +110,42 @@ def add_text(self, tag, text, step=None):
                             step=step)
     
     @ignore_exception
-    def display_attention_heads(self, outputs, tag='', step=None):
+    def display_attention_heads(self, outputs, tag='', step=None, fname=None):
```
Comment (cschaefer26) [inline utils/logging_utils.py:113]: Type hinting would be really helpful here.

File: utils/metrics.py [hunk 2]
```diff
@@ -20,7 +19,7 @@ def attention_score(att, mel_len, phon_len, r):
     peak_score = attention_peak_score(att, mask)
     
     # diagonality
-    diag_score = diagonality_measure(att, mel_len, phon_len)
+    diag_score = diagonality_score(att, mel_len, phon_len)
     
     return loc_score, peak_score, 3. / diag_score
```
Comment (cschaefer26) [inline utils/metrics.py:24]: Why the 3 here?

**Review by cschaefer26 — COMMENTED** [review]
Generally good changes, I am just a bit unsure if its really necessary to have the extra breathing logic. Imo breathing should just be a special case of a standard model (maybe include a start token anyways).

File: create_training_data.py [hunk 3]
```diff
@@ -117,14 +151,14 @@ def process_phonemes(file_id):
...
     for (file_id, phonemes) in phon_iter:
         phonemized_data.update({file_id: phonemes})
     
     print('\nPhonemized metadata samples:')
     for i in sample_items:
         print(f'{i}:{phonemized_data[i]}')
-        summary_manager.add_text(f'Phonemized samples/{i}', text=phonemized_data[i])
+        summary_manager.add_text(f'{i}/phonemes', text=phonemized_data[i])
     
     new_metadata = [''.join([key, '|', phonemized_data[key], '\n']) for key in phonemized_data]
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply create_training_data.py:163]: Cleaner, would be missing the '\n' though, right?
```suggestion
    new_metadata = [f'{k}|{v}\n' for k, v in phonemized_data.items()]
```

File: data/datasets.py [hunk 4]
```diff
@@ -27,39 +27,83 @@ class DataReader:
     training data.
     """
     
-    def __init__(self, data_directory: str, metadata_path: str, metadata_reading_function=None, scan_wavs=False):
+    def __init__(self, wav_directory: str, metadata_path: str, metadata_reading_function=None, scan_wavs=False,
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply data/datasets.py:30]: Yeah a little convoluted. It stems from the need of distinguishing between processed metadata and original metadata. Due to some legacy code I didn't wanna change yet :D The filename list seems needed to me due to the upsampling addition. oder?

File: data/text/symbols.py [hunk 5]
```diff
@@ -6,9 +6,7 @@
 _diacrilics = 'ɚ˞ɫ'
 _phonemes = sorted(list(
     _vowels + _non_pulmonic_consonants + _pulmonic_consonants + _suprasegmentals + _other_symbols + _diacrilics))
-_punctuations = '!,-.:;? '
+_punctuations = '!,-.:;? \'()'
 _alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzäüöß'
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply data/text/symbols.py:10]: True, although it's actually only to detect sentences that are particularly short
```suggestion
_alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzäüößÄÖÜ'
```

File: data/text/tokenizer.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        self.add_start_end = add_start_end
+        if add_start_end:
+            self.start_token_index = len(self.alphabet) + 1
+            self.end_token_index = len(self.alphabet) + 2
+            self.vocab_size += 2
+            self.idx_to_token[self.start_token_index] = start_token
+            self.idx_to_token[self.end_token_index] = end_token
+        self.model_breathing = model_breathing
+        if model_breathing:
+            self.initial_breathing_token_index = self.vocab_size
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply data/text/tokenizer.py:30]: Nope, but maybe then it's not well written.

File: extract_durations.py [hunk 7]
```diff
@@ -1,148 +1,138 @@
...
+                                               scalar_value=tf.reduce_mean(batch_avg_diag_measure[i]), step=c)
+            
+            for i, name in enumerate(file_name_batch):
+                all_durations = np.append(all_durations, durations[i])  # for plotting only
+                summary_manager.add_image(tag='ExtractedAlignments',
+                                          image=tf.expand_dims(tf.expand_dims(final_align[i], 0), -1),
+                                          step=step)
+                
+                step += 1
+                np.save(str(target_dir / f"{name.numpy().decode('utf-8')}.npy"), durations[i])
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply extract_durations.py:102]: gotta love double quotes

File: utils/metrics.py (hunk 2 above)
Comment (cfrancesco) [reply utils/metrics.py:24]: To bring the values in the same range

File: utils/logging_utils.py (hunk 1 above)
Comment (cfrancesco) [reply utils/logging_utils.py:113]:
```suggestion
    def display_attention_heads(self, outputs: dict, tag='', step=None, fname=None):
```

File: predict_tts.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    elif args.text is not None:
+        text = [args.text]
+        fname = 'custom_text'
+    else:
+        fname = None
+        text = None
+        print(f'Specify either an input text (-t "some text") or a text input file (-f /path/to/file.txt)')
+        exit()
+    config_loader = Config(config_path=args.config)
+    if args.outdir is None:
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply predict_tts.py:35]: Should include line below

File: predict_tts.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        outdir = config_loader.log_dir
```
Comment (cfrancesco) [inline predict_tts.py:36]:
```suggestion
    outdir = Path(args.outdir) if args.outdir is not None else config_loader.log_dir
```

File: train_aligner.py [hunk 10]
```diff
@@ -159,32 +185,11 @@ def validate(model,
...
     
-    if model.step % config['validation_frequency'] == 0:
+    if model.step % config['validation_frequency'] == 0 and (model.step >= config['prediction_start_step']):
         val_loss, time_taken = validate(model=model,
                                         val_dataset=valid_dataset,
-                                        summary_manager=summary_manager)
+                                        summary_manager=summary_manager,
+                                        weighted_durations=config['extract_attention_weighted'])
         t.display(f'validation loss at step {model.step}: {val_loss} (took {time_taken}s)',
                   pos=len(config['n_steps_avg_losses']) + 3)
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply train_aligner.py:194]: To separate it from the rest while printing.

File: data/metadata_readers.py [hunk 11]
```diff
@@ -31,6 +32,26 @@ def ljspeech(metadata_path: str, column_sep='|') -> dict:
...
+    """
+    Used to read metadata files created within the repo.
+    """
+    text_dict = {}
+    upsample = []
+    with open(metadata_path, 'r', encoding='utf-8') as f:
+        for l in f.readlines():
+            l_split = l.split(column_sep)
+            filename, text = l_split[0], l_split[1]
+            if filename.endswith('.wav'):
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply data/metadata_readers.py:46]: True. Legacy is hard to purge

File: data/metadata_readers.py [hunk 12]
```diff
@@ -31,6 +32,26 @@ def ljspeech(metadata_path: str, column_sep='|') -> dict:
...
+                filename = filename.split('.')[0]
```
Comment (cfrancesco) [inline data/metadata_readers.py:47]:
```suggestion
```

File: config/tts_config.yaml [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+# TRAINING
+dropout_rate: 0.1
+learning_rate_schedule:
+  - [0, 1.0e-4]
+max_steps: 260_000
+debug: False
+
+# LOGGING
+validation_frequency: 5_000
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply config/tts_config.yaml:27]: besserwisser

File: extract_durations.py [hunk 14]
```diff
@@ -1,148 +1,138 @@
...
+            pitch = np.load((config_manager.pitch_dir / sample_name).with_suffix('.npy').as_posix())
+            durations = np.load((config_manager.duration_dir / sample_name).with_suffix('.npy').as_posix())
+            mel = np.load((config_manager.mel_dir / sample_name).with_suffix('.npy').as_posix())
+            # text = metadatareader.text_dict[sample_name]
+            char_wise_pitch = _pitch_per_char(pitch, durations, mel.shape[0])
+            # len_text = len(text)
+            # if config_manager.config['model_breathing']:
+            #     len_text += 1 # TODO: improve this, BREATHING TOKEN needs +1
+            # assert char_wise_pitch.shape[0] == len_text, \
+            #     f'{sample_name}: dshape {char_wise_pitch.shape} == tshape {len_text}'
```
Comment (cfrancesco) [inline extract_durations.py:129]:
```suggestion
```

Comment (cfrancesco) [inline extract_durations.py:129]:
```suggestion
            char_wise_pitch = _pitch_per_char(pitch, durations, mel.shape[0])
```

File: README.md [hunk 15]
```diff
@@ -123,28 +125,30 @@ python train_forward.py --config config/melgan
...
+python predict_tts.py -t "Please, say something." --config config/session_paths.yaml
+```
+Or in a python script
 ```python
 from utils.config_manager import Config
-from utils.audio import Audio
+from data.audio import Audio
 
-config_loader = Config(config_path=f'/path/to/config/', model_kind=f'forward')
+config_loader = Config(config_path=f'config/session_paths.yaml', tts=True)
```
Comment (cfrancesco) [inline README.md:144]:
```suggestion
config_loader = Config(config_path=f'config/session_paths.yaml')
```
